miner/Miner.hs

```diff
@@ -0,0 +1,285 @@
+-- work is submitted.
+--
+
+module Main ( main ) where
+
+import BasePrelude hiding (app, option)
+import Control.Concurrent.Async (async, race, wait)
+import Control.Concurrent.STM.TMVar
+import Control.Concurrent.STM.TVar (modifyTVar')
+import Control.Error.Util (note)
+import Control.Scheduler (Comp(..), replicateWork, terminateWith, withScheduler)
+import Crypto.Hash.Algorithms (SHA512t_256)
+import Crypto.Hash.IO
+import qualified Data.ByteArray as BA
+import Data.Bytes.Put
+import qualified Data.ByteString as B
+import Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HM
+import qualified Data.Text as T
+import Data.Tuple.Strict (T2(..))
+import Foreign.Marshal.Alloc (allocaBytes)
+import qualified Network.Wai.Handler.Warp as W
+import Options.Applicative
+import Servant.API
+import Servant.Server (Application, Server, serve)
+import Servant.Swagger (toSwagger)
+import Servant.Swagger.UI (SwaggerSchemaUI, swaggerSchemaUIServer)
+
+-- internal modules
+import Chainweb.BlockHeader
+import Chainweb.Difficulty (HashTarget, encodeHashTarget)
+import Chainweb.RestAPI.Orphans ()
+import Chainweb.Time (Micros, Time, encodeTimeToWord64, getCurrentTimeIntegral)
+import Chainweb.Utils (int, runGet)
+import Chainweb.Version (ChainId, ChainwebVersion(..), chainwebVersionFromText)
+import Miner.Types ()
+
+---
+
+--------------------------------------------------------------------------------
+-- Servant
+
+type MinerAPI = "submit" :> ReqBody '[JSON] BlockHeader :> Post '[JSON] ()
+    :<|> "poll" :> Capture "chainid" ChainId
+                :> Capture "blockheight" BlockHeight
+                :> Get '[JSON] (Maybe BlockHeader)
+
+type API = MinerAPI :<|> SwaggerSchemaUI "help" "swagger.json"
+
+server :: Env -> Server API
+server e = (liftIO . submit e :<|> (\cid h -> liftIO $ poll e cid h))
+    :<|> swaggerSchemaUIServer (toSwagger (Proxy :: Proxy MinerAPI))
+
+app :: Env -> Application
+app = serve (Proxy :: Proxy API) . server
+
+--------------------------------------------------------------------------------
+-- CLI
+
+type ResultMap = HashMap (T2 ChainId BlockHeight) BlockHeader
+
+data Env = Env
+    { work :: TMVar BlockHeader
+    , results :: TVar ResultMap
+    , cores :: Word16
+    , version :: ChainwebVersion
+    , port :: Int
+    }
+
+pEnv :: TMVar BlockHeader -> TVar ResultMap -> Parser Env
+pEnv tbh thm = Env tbh thm <$> pCores <*> pVersion <*> pPort
+
+pCores :: Parser Word16
+pCores = option auto
+    (long "cores" <> metavar "COUNT" <> value 1
+     <> help "Number of CPU cores to use (default: 1)")
+
+pVersion :: Parser ChainwebVersion
+pVersion = option cver
+    (long "version" <> metavar "VERSION"
+     <> value Testnet01
+     <> help "Chainweb Network Version (default: testnet01)")
+  where
+    cver :: ReadM ChainwebVersion
+    cver = eitherReader $ \s ->
+        note "Illegal ChainwebVersion" . chainwebVersionFromText $ T.pack s
+
+pPort :: Parser Int
+pPort = option auto
```

Comment:
Not all `Int` are valid ports.

You'll also want to include the interface to listen on. In production, opften that will be only a local network or a VPN.
f
I propose using the exiting types and configuration tools parsers and that we use already in chainweb to keep the configuration UIs for all components consistent.